Forwarded: not-needed

--- a/options.h	2013-05-20 15:49:18.000000000 +0800
+++ b/options.h	2013-05-20 15:49:27.000000000 +0800
@@ -197,7 +197,7 @@
  * note that it will be provided for all "hidden" client-interactive
  * style prompts - if you want something more sophisticated, use 
  * SSH_ASKPASS instead. Comment out this var to remove this functionality.*/
-/*#define DROPBEAR_PASSWORD_ENV "DROPBEAR_PASSWORD"*/
+#define DROPBEAR_PASSWORD_ENV "DROPBEAR_PASSWORD"
 
 /* Define this (as well as ENABLE_CLI_PASSWORD_AUTH) to allow the use of
  * a helper program for the ssh client. The helper program should be
